### DISCUSSION TITLE
Deployment Error with Google Cloud

### DISCUSSION THREAD

**Discussion by akf0rce** [vikejs/vike#1198] [Discussion]
### Description

Hello Vike team;

I'm trying to deploy to Google Cloud (AppEngine);

Config;

```
const config: UserConfig = {
  plugins: [react(), ssr()],
  ssr: {
    noExternal,
  },
  resolve: {
    alias: {
      '#root': __dirname,
    },
  },
}

export default config
```

after running the build and creating an `app.yaml` file;

```
runtime: nodejs14 # or your desired Node.js version
service: dev

instance_class: F2 # choose an appropriate instance class; F2 is just an example

automatic_scaling:
  target_cpu_utilization: 0.65 # adjust as needed for your application's performance

entrypoint: node dist/server/importBuild.cjs
```

I get the following error on Google Cloud Logs;

```
"Error: Cannot find module '/workspace/dist/server/importBuild.cjs'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:931:15)
    at Function.Module._load (internal/modules/cjs/loader.js:774:27)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:75:12)
    at internal/main/run_main_module.js:17:47 {"
```

package.json

```
  "scripts": {
    "dev": "npm run server:dev",
    "prod": "npm run lint && npm run build && npm run server:prod",
    "build": "vite build",
    "server": "node --loader ts-node/esm ./server/index.ts",
    "server:dev": "npm run server",
    "server:prod": "cross-env NODE_ENV=production npm run server",
    "lint": "eslint . --max-warnings 0"
  },
```
    
Could I get some guidance, please? 🙏🏽

**Comment by brillout** [comment]
Is this reproducible locally? So that I can have a quick look at it.

**Comment by akf0rce** [comment]
Everything works as intended locally it's more so when trying to deploy to appEngine via gCloud.

```
npm run build
cp app.yaml ./dist
gcloud app deploy ./dist
```

https://cloud.google.com/appengine/docs/standard/nodejs/runtime#application_startup

**Comment by brillout** [comment]
Doesn't App Engine have a preview thing to simulate a deployment locally? (Been ages last time I used it.)

**Comment by akf0rce** [comment]
It does, but not for node instances;

https://cloud.google.com/appengine/docs/standard/tools/using-local-server?tab=node.js#top

I've managed to get past the first issue and now I'm presented with the following:

```
(node:11) UnhandledPromiseRejectionWarning: Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/node_modules/vike/dist/esm/node/runtime/globalContext/loadImportBuild.js' imported from /workspace/server/importBuild.cjs
    at new NodeError (internal/errors.js:322:7)
    at finalizeResolution (internal/modules/esm/resolve.js:318:11)
    at moduleResolve (internal/modules/esm/resolve.js:776:10)
    at Loader.defaultResolve [as _resolve] (internal/modules/esm/resolve.js:887:11)
    at Loader.resolve (internal/modules/esm/loader.js:89:40)
    at Loader.getModuleJob (internal/modules/esm/loader.js:242:28)
    at Loader.import (internal/modules/esm/loader.js:177:28)
    at importModuleDynamically (internal/modules/cjs/loader.js:1056:27)
    at exports.importModuleDynamicallyCallback (internal/process/esm_loader.js:30:14)
    at /workspace/server/importBuild.cjs:3:37
(Use `node --trace-warnings ...` to show where the warning was created)
(node:11) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
Exit 0 app (pid 11): /bin/sh -c exec node server/importBuild.cjs
```

**Comment by brillout** [comment]
It seems that App Engine's Node.js runtime is doing something unusual.

What's the content of the file `/workspace/server/importBuild.cjs`?

Does the path `/node_modules/vike/dist/esm/node/runtime/globalContext/loadImportBuild.js` exist?

Looking forward for us to create a working example.

**Comment by akf0rce** [comment]
```
// File generated by https://github.com/brillout/vite-plugin-import-build
(async () => {
  const { setImportBuildGetters } = await import('../../node_modules/vike/dist/esm/node/runtime/globalContext/loadImportBuild.js');
  setImportBuildGetters({
    pageFiles: () => import('./pageFiles.mjs'),
    clientManifest: () => require('../assets.json'),
    pluginManifest: () => require('../client/vike.json'),
  });
})()
```
I think this is the culprit here - according to this line: `import('../../node_modules/vike/dist/esm/node/runtime/globalContext/loadImportBuild.js');` this script is trying to import the `node_modules` from the root and not the `./dist` directory;

Is it intentional to not have the node_modules included in the build directory?

To answer your question, the `node_modules` is not deployed to the server as it lives outside the `./dist` directory.

In previous `React` + `Vite` builds, the `node_modules` are transposed into an `index.js` available within the build directory.

**Comment by brillout** [comment]
In production, where does `node_modules/` live?

**Comment by akf0rce** [comment]
I only deployed the `./dist` directory as I would with `vite` projects, locally on my machine, it's in the project root with all the dev files and on the production server (appEngine), it doesn't exist as it's not bundled within the `build dir`

**Comment by brillout** [comment]
Is that what you *want* or is it an App Engine requirement?

**Comment by akf0rce** [comment]
It is only my intention to upload production files; So, let's break this down;

Let's say I'm working in a directory called `vike-app`, and within that app I have the usual folder structure as one would when you set up a new scaffold.

Then I run `npm run build`, which outputs to `./dist`.

I then deploy the output of the `./dist` directory to AppEngine using `gcloud deploy app ./dist` as the development files are not required in production.

The structure of the `build` directory

```
| - /dist
| -- /client
| -- /server
```

In the file `server/importBuild.cjs` line 3 of the script reads;

`const { setImportBuildGetters } = await import('../../node_modules/vike/dist/esm/node/runtime/globalContext/loadImportBuild.js');`

As the `importBuild.cjs` script is looking for the `node_modules` from outside the `./dist` directory, then is it `Vike's` intention to also upload the `node_modules` directory to the server?

After looking at the `./dist` directory from just a standard `Vite` build `node_modules` are bundled into an `index-xxxxxxx.js` file, which is included in the `/assets` directory within the build dir `./dist`

**Comment by brillout** [comment]
I understand all that but still: can you run `$ npm install` for production so that `node_modules/` exists?

I know that isn't what you want, but that would work right?

As for self-contained distributions for production, there are a couple of ways to make it work. (FYI it's on the radar and high priority to improve the DX around self-contained distributions.)

First things first, I'd suggest we make the App Engine deployment work. Once it works we can then focus on having a self-contained distribution.

**Comment by akf0rce** [comment]
What would be the steps to run npm install for production?

**Comment by brillout** [comment]
That's an App Engine question.

**Comment by brillout** [comment]
Moving to GitHub Discussions. Let's continue the conversation over there.

**Comment by kochrt** [comment]
I want to bump this; though I have a slightly different issue that I think is related. 


I'm [working on a static site generator](https://github.com/mark-when/reblog) (like hugo/gatsby/jekyll) that I'm using vike for. I wanted to use the SSG/prerender aspect of vike to build the site scaffold, distribute it as some kind of executable or CLI, then an end user would provide their input file (a [markwhen](https://docs.markwhen.com) document) and the generator would spit out a complete multi-page static site. I made a [simple script](https://github.com/mark-when/reblog/blob/main/lib/index.js) that just calls `prerender` and I wanted end users to be able to call that with their markwhen file set as an environment variable so it would build the site based on their provided file.


However I'm hitting the same snag as above - it's expecting the `loadImportBuild.js` stuff. I tried explicitly adding `node_modules` to the package I'm distributing to `npm` but it is ignored.


Should I even be using vike? It may just be the wrong tool. If you do end up looking at the [repo](https://github.com/mark-when/reblog) @brillout I apologize for how badly it's set up right now... I'm just trying to get something that works.


Sorry for opening and bumping a bunch of random issues, I know you have a lot on your plate. Normally I'm able to put something together but I've kind of hit a wall here.

**Reply by brillout** [reply]
I don't usually look at user's projects (it's quite time consuming), that said minimal reproduction welcome and I'll have a look at it.

**Comment by kochrt** [comment]
Continuing: I've tried `npm pack`ing, manually trying to add `node_modules/vike`, but have so far been unsuccessful. I've thought about requiring end users to install `vike` and then copy it into their `node_modules/reblog/node_modules/vike` so that the generated code 
```  
const { setImportBuildGetters } = await import('../../node_modules/vike/dist/esm/node/runtime/globalContext/loadImportBuild.js');
``` 
works. But this all seems very onerous and I haven't even gotten it working myself yet. 

I've tried adding `vike` as a `bundledDependency` but then I get a `415 Unsupported Media Type - PUT https://registry.npmjs.org/mwblog - Hard link is not allowed` when trying to publish to npm.

**Reply by brillout** [reply]
I ain't sure exactly what you to achieve, but if you want to bundle the server-side then consider using esbuild (that's what tools like wrangler do).

FYI I'm going to work on https://github.com/vikejs/vike/issues/1165 and https://github.com/vikejs/vike/issues/562 after releasing the V1 Design. So now would be a good time to dig into what exactly it is you need from Vike.
